src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterTxSource.cs

```diff
@@ -0,0 +1,333 @@
+
+        if (CheckDecryptionKeys(decryptionKeys, eonInfo.Value))
+        {
+            if (_logger.IsInfo) _logger.Info($"Validated Shutter decryption key for slot {decryptionKeys.Gnosis.Slot}");
+
+            List<SequencedTransaction> sequencedTransactions = GetNextTransactions(decryptionKeys.Eon, decryptionKeys.Gnosis.TxPointer);
+            if (_logger.IsInfo) _logger.Info($"Got {sequencedTransactions.Count} transactions from Shutter mempool...");
+
+            _loadedTransactions = DecryptSequencedTransactions(sequencedTransactions, decryptionKeys);
+            _loadedTransactionsSlot = decryptionKeys.Gnosis.Slot;
+
+            if (_logger.IsInfo)
+            {
+                _loadedTransactions.ForEach(tx => _logger.Info(tx.ToShortString()));
+            }
+        }
+    }
+
+    private ulong GetNextSlot()
+    {
+        // assume Gnosis or Chiado chain
+        ulong genesisTimestamp = (_specProvider.ChainId == BlockchainIds.Chiado) ? ChiadoSpecProvider.BeaconChainGenesisTimestamp : GnosisSpecProvider.BeaconChainGenesisTimestamp;
+        return (((ulong)DateTimeOffset.UtcNow.ToUnixTimeSeconds() - genesisTimestamp) / 5) + 1;
+    }
+
+    private IEnumerable<Transaction> DecryptSequencedTransactions(IEnumerable<SequencedTransaction> sequencedTransactions, Dto.DecryptionKeys decryptionKeys)
+    {
+        // order by identity preimage to match decryption keys
+        IEnumerable<(int, Transaction?)> unorderedTransactions = sequencedTransactions
+            .Select((x, index) => x with { Index = index })
+            .OrderBy(x => x.IdentityPreimage, Bytes.Comparer)
+            .Zip(decryptionKeys.Keys.Skip(1))
+            .Select(x => (x.Item1.Index, DecryptSequencedTransaction(x.Item1, x.Item2)));
+
+        // return decrypted transactions to original order
+        IEnumerable<Transaction> transactions = unorderedTransactions
+            .OrderBy(x => x.Item1)
+            .Select(x => x.Item2)
+            .OfType<Transaction>();
+
+        return transactions;
+    }
+
+    private bool CheckDecryptionKeys(in Dto.DecryptionKeys decryptionKeys, in ShutterEon.Info eonInfo)
+    {
+        if (decryptionKeys.InstanceID != _instanceId)
+        {
+            if (_logger.IsDebug) _logger.Debug($"Invalid decryption keys received on P2P network: instanceID {decryptionKeys.InstanceID} did not match expected value {_instanceId}.");
+            return false;
+        }
+
+        if (decryptionKeys.Eon != eonInfo.Eon)
+        {
+            if (_logger.IsDebug) _logger.Debug($"Invalid decryption keys received on P2P network: eon {decryptionKeys.Eon} did not match expected value {eonInfo.Eon}.");
+            return false;
+        }
+
+        foreach (Dto.Key key in decryptionKeys.Keys.AsEnumerable().Skip(1))
```

Comment:
why Skip(1)?

Reply:
The first decryption key is a placeholder that needs to be skipped

Reply:
1. Comment
2. You do this skip in multiple places, should be done in only one

Reply:
I added comments but there is one place where all of the decryption keys (including the placeholder) are needed so I think it would be clunkier to do in one place